plugins/woocommerce/src/Internal/Admin/Orders/Edit.php

```diff
@@ -0,0 +1,202 @@
+		add_meta_box( 'woocommerce-order-downloads', __( 'Downloadable product permissions', 'woocommerce' ) . wc_help_tip( __( 'Note: Permissions for order items will automatically be granted when the order status changes to processing/completed.', 'woocommerce' ) ), 'WC_Meta_Box_Order_Downloads::output', $screen_id, 'normal', 'default' );
+		/* Translators: %s order type name. */
+		add_meta_box( 'woocommerce-order-actions', sprintf( __( '%s actions', 'woocommerce' ), $title ), 'WC_Meta_Box_Order_Actions::output', $screen_id, 'side', 'high' );
+	}
+
+	/**
+	 * Enqueue necessary scripts for order edit page.
+	 */
+	private function enqueue_scripts() {
+		if ( wp_is_mobile() ) {
+			wp_enqueue_script( 'jquery-touch-punch' );
+		}
+		wp_enqueue_script( 'post' ); // Ensure existing JS libraries are still available for backward compat.
+	}
+
+	/**
+	 * Setup hooks, actions and variables needed to render order edit page.
+	 */
+	private function setup() {
+		$this->initialize_order_object();
+		$current_screen = get_current_screen();
+		$current_screen->is_block_editor( false );
+		$this->screen_id = $current_screen->id;
+		$this->add_order_meta_boxes( $this->screen_id, __( 'Order', 'woocommerce' ) );
+		$this->add_order_specific_meta_box();
+
+		/**
+		 * From wp-admin/includes/meta-boxes.php.
+		 *
+		 * Fires after all built-in meta boxes have been added. Custom metaboxes may be enqueued here.
+		 *
+		 * @since 3.8.0.
+		 */
+		do_action( 'add_meta_boxes', wc_get_page_screen_id( 'shop-order' ), $this->order );
+		$this->enqueue_scripts();
+	}
+
+	/**
+	 * Setup order object to be used in metaboxes.
+	 *
+	 * @return void
+	 */
+	private function initialize_order_object() {
+		global $theorder;
+		$action = sanitize_text_field( wp_unslash( isset( $_GET['action'] ) ? $_GET['action'] : '' ) );
+		switch ( $action ) {
+			case 'edit':
+				$this->order = wc_get_order( absint( isset( $_GET['id'] ) ? $_GET['id'] : 0 ) );
+				$this->verify_edit_object_and_permission();
+				break;
+			case 'new':
+				$this->order = new \WC_Order();
+				if ( ! current_user_can( 'publish_shop_orders' ) && ! current_user_can( 'manage_woocommerce' ) ) {
+					wp_die( esc_html__( 'You don\'t have permission to create a new order', 'woocommerce' ) );
+				}
+				break;
+			default:
+				wp_redirect( admin_url( 'admin.php?page=wc-orders' ) );
+				exit;
+		}
+		$theorder = $this->order;
+	}
+
+	/**
+	 * Verify that the order object is valid and user has permission to edit it.
+	 */
+	private function verify_edit_object_and_permission() {
+		if ( ! isset( $this->order ) || ! $this->order ) {
+			wp_die( esc_html__( 'You attempted to edit an item that does not exist. Perhaps it was deleted?', 'woocommerce' ) );
+		}
+
+		if ( ! current_user_can( 'edit_others_shop_orders' ) && ! current_user_can( 'manage_woocommerce' ) ) {
+			wp_die( esc_html__( 'You do not have permission to edit this order', 'woocommerce' ) );
```

Comment:
**Not a blocker.** *Though I wanted to flag it, it works as is (so we can move on if we prefer).*

Using `wp_die()` at this stage (because other bits of the editor UI have already been rendered) looks visually less polished than what we see with the regular post editor. Our editor:

<img width="649" alt="wp-die-late" src="https://user-images.githubusercontent.com/3594411/177428873-bac07748-eb38-4e56-8d2a-8c2f9c718fa0.png">

Vs the post editor:

<img width="840" alt="wp-die-early" src="https://user-images.githubusercontent.com/3594411/177428894-ccedc3ec-1e7b-40ee-8a6b-6e0b912abdc6.png">

If we wished, we could probably move this check further 'back in time' to get a consistent result.

Reply:
Not trying to hijack the review or anything, just wanted to mention that integrating the `Edit` class with the `PageController` from #33687 ([link](https://github.com/woocommerce/woocommerce/blob/0a2588f6c41355417fe4744c25362aa4eab0ff46/plugins/woocommerce/src/Internal/Admin/Orders/PageController.php)) will probably help with this: once the actions here (new/edit) are correctly configured, the setup function will be called before anything is rendered (and thus the error will work as in the post editor). This is on purpose in case we need to perform redirects or stuff like that, which need to happen at an earlier stage.

In any case, there are some conflicts with `trunk` and my suggestion above might be a way to address those 🙃. It'd also be a good opportunity to both re-use and make any changes the `PageController` might need.